java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/LspTemplateUI.java

```diff
@@ -0,0 +1,265 @@
+                }
+            } catch (MalformedURLException ex) {
+                continue;
+            }
+        }
+        String root = System.getProperty("user.home"); // NOI18N
+        return DataFolder.findFolder(FileUtil.toFileObject(new File(root)));
+    }
+
+    private static String singleSelection(List<QuickPickItem> selectedGroups) {
+        if (selectedGroups == null || selectedGroups.size() != 1) {
+            throw raise(RuntimeException.class, new UserCancelException(""));
+        }
+        return selectedGroups.get(0).getUserData().toString();
+    }
+
+    private static DataFolder findTargetFolder(ExecuteCommandParams params, List<WorkspaceFolder> folders, boolean[] defaultTargetExact) {
+        for (Object arg : params.getArguments()) {
+            if (arg == null) {
+                continue;
+            }
+            String path;
+            if (arg instanceof JsonPrimitive) {
+                JsonPrimitive jp = (JsonPrimitive) arg;
+                path = jp.getAsString();
+            } else {
+                path = arg.toString();
+            }
+            try {
+                FileObject fo = URLMapper.findFileObject(new URL(path));
+                for (;;) {
+                    if (fo == null) {
+                        break;
+                    }
+                    if (!fo.isFolder()) {
+                        fo = fo.getParent();
+                        defaultTargetExact[0] = false;
+                        continue;
+                    }
+                    return DataFolder.findFolder(fo);
+                }
+            } catch (MalformedURLException ex) {
+                continue;
+            }
+        }
+        defaultTargetExact[0] = false;
+        return findWorkspaceRoot(folders);
+    }
+
+    private static String removeExtensionFromFileName(String nameWithExtension, String templateExtension) {
+        if (nameWithExtension.endsWith('.' + templateExtension)) {
+            return nameWithExtension.substring(0, nameWithExtension.length() - templateExtension.length() - 1);
+        } else {
+            return nameWithExtension;
+        }
+    }
+
+    private static List<QuickPickItem> quickPickTemplates(final DataFolder folder) {
+        Node[] arr = folder.getNodeDelegate().getChildren().getNodes(true);
+        List<QuickPickItem> categories = new ArrayList<>();
+        for (Node n : arr) {
+            FileObject fo = n.getLookup().lookup(FileObject.class);
+            DataObject obj = n.getLookup().lookup(DataObject.class);
+
+            boolean display;
+            if (obj instanceof DataFolder) {
+                Object simple = fo.getAttribute("simple"); // NOI18N
+                if (simple != null) {
+                    display = Boolean.TRUE.equals(simple);
+                } else {
+                    Object cathegory = fo.getAttribute("templateCategory"); // NOI18N
+                    if ("invisible".equals(cathegory)) { // NOI18N
```

Comment:
Are these tags documented ?

Reply:
Tested in `ide/projectui/test/unit/src/org/netbeans/modules/project/ui/TemplateChooserPanelGUITest.java` and documented in [Project UI javadoc](https://bits.netbeans.org/12.3/javadoc/org-netbeans-modules-projectuiapi/org/netbeans/spi/project/ui/templates/support/package-summary.html) - however there is no mention of these properties in `arch.xml` as far as I can tell.

Reply:
The API is documented in architecture questions document since 904da56. Thanks for encouraging proper API documentation policy!